### DISCUSSION TITLE
Missing milk TP or Lactose can't be parsed to calculate_Trg_NEmilk_Milk

### DISCUSSION THREAD

**Discussion by davos-i** [CNM-University-of-Guelph/NASEM-Model-Python#144] [NASEM 8 R/Book]
Currently it is not possible to import missing values for milk TP or Lactose, but this function has option to calculate milk NE if they are missing.
Maybe we use 0 if missing, not None. That way the `animal_input` dict will always contain all 3 of these (and is still stored as a float), but can be 0 if unknown. This would also prevent 0's being used if accidentally entered here (e.g. from automatic records)
```
animal_input['Trg_MilkFatp']
animal_input['Trg_MilkTPp']
animal_input['Trg_MilkLacp']
```

[ ] Document how to parse values to trigger both equations
[ ] Add tests for all conditions

https://github.com/CNM-University-of-Guelph-dev/NASEM-Model-Python/blob/caaf2808761c97da870ff3668baa7cb3dcf562b1/src/nasem_dairy/NASEM_equations/dev_milk_equations.py#L6-L60

**Comment by BFieguth** [comment]
The input_validation module makes it impossible for Trg_MilkFatp, Trg_MilkTPp or Trg_MilkLacp to be None. execute_model will throw an error during validation if they do. The following line is included for completeness but should never be used unless someone calls calculate_Trg_NEmilk_Milk directly. 

`   
 if Trg_MilkLacp is None or Trg_MilkTPp is None:
        Trg_NEmilk_Milk = 0.36 + 9.69 * Trg_MilkFatp/100    # Line 384/2887
`

I don't think any changes should be made to the function. If we are worried about users inputing invalid values this should be incorporated into the input_validation module. For example, we know milk fat, protein and lactose will not be 0%, therfore we can check for this during validation and throw a warning or error.

**Comment by davos-i** [comment]
I think this was added so that if you don't know, or don't want to use, a target milk TP or Lactose then the model can still calculate a target NE milk to use for the energy requirements. So I think we still need to require the user to be intentional (i.e. not allow a 'none') which is done by input validation, but we could make it so that if they set a target of 0% for either TP or Lactose then it means they are basically choosing to use this equation?

**Comment by BFieguth** [comment]
I see what you mean. Trg_MilkLacp is only used in 2 equations (the other just calculates the kg in milk) so it would be easy to add some additional logic around it. The issue is with Trg_MilkTPp as it is used in several equations and some of them have no checks for values of 0/None.

Perhaps allowing values of 0 for Trg_MilkTPp and Trg_MilkLacp is not the right approach. The model does support milk protein predictions using mPrt_eqn. If a user does not know the milk protein then is it better to "ignore" it (by setting to 0) or use a predicted value? As for Trg_MilkLacp we know this does not change as much as protein or fat. Is it better to use the default of 4.85 than to use 0?
